django/db/models/fields/composite.py

```diff
@@ -0,0 +1,146 @@
+
+        cols = self.lhs.get_source_expressions()
+        exprs = [Exact(col, val) for col, val in zip(cols, self.rhs)]
+
+        return compiler.compile(WhereNode(exprs, connector=AND))
+
+
+class TupleIn(In):
+    def get_prep_lookup(self):
+        if not isinstance(self.lhs, MultiColSource):
+            raise ValueError(
+                "The left-hand side of the 'in' lookup must be an instance of "
+                "MultiColSource"
+            )
+        if not isinstance(self.rhs, (tuple, list)):
+            raise ValueError(f"'{self.lhs.field.name}' must be a tuple or a list")
+        if not all(isinstance(vals, (tuple, list)) for vals in self.rhs):
+            raise ValueError(
+                f"'{self.lhs.field.name}' must be a collection of tuples or lists"
+            )
+        if not all(len(self.lhs) == len(vals) for vals in self.rhs):
+            raise ValueError(
+                f"'{self.lhs.field.name}' must have {len(self.lhs)} elements"
+            )
+
+        return super().get_prep_lookup()
+
+    def as_sql(self, compiler, connection):
+        from django.db.models.sql.where import AND, OR, WhereNode
+
+        exprs = []
+        cols = self.lhs.get_source_expressions()
+
+        for vals in self.rhs:
+            exprs.append(
+                WhereNode(
+                    [Exact(col, val) for col, val in zip(cols, vals)], connector=AND
+                )
+            )
+
+        return compiler.compile(WhereNode(exprs, connector=OR))
+
+
+class CompositeValue(tuple):
+    def is_set(self):
+        return all(value is not None for value in self)
+
+
+class CompositeAttribute:
+    def __init__(self, field):
+        self.field = field
+
+    def __get__(self, instance, cls=None):
+        return CompositeValue(
+            getattr(instance, field_name) for field_name in self.field.field_names
+        )
+
+    def __set__(self, instance, values):
+        if values is None:
+            values = (None,) * len(self.field.field_names)
+
+        for field_name, value in zip(self.field.field_names, values):
+            setattr(instance, field_name, value)
+
+
+class CompositePrimaryKey(Field):
+    descriptor_class = CompositeAttribute
+
+    def __init__(self, *args, **kwargs):
+        kwargs["db_column"] = None
+        kwargs["editable"] = False
+        super().__init__(**kwargs)
+        self.field_names = args
+        self.fields = None
+
+    def contribute_to_class(self, cls, name, **_):
+        super().contribute_to_class(cls, name, private_only=True)
+        cls._meta.pk = self
+        setattr(cls, self.attname, self.descriptor_class(self))
+
+    def get_attname_column(self):
+        return self.get_attname(), self.db_column
+
+    def __iter__(self):
+        return iter(self.fields)
+
+    def __len__(self):
+        return len(self.fields)
+
+    @cached_property
+    def cached_col(self):
+        return MultiColSource(self.model._meta.db_table, self.fields, self.fields, self)
```

Comment:
The fact that you set `self.fields` for both `source` and `target` is a good sign that `MultiColSource` is not the appropriate expression for `cached_col`.

Reply:
I think adding a new expression `Cols` makes most sense